spark/dl/src/main/scala/com/intel/analytics/bigdl/parameters/ParameterOperations.scala

```diff
@@ -0,0 +1,135 @@
+import com.intel.analytics.bigdl.utils.Table
+import org.apache.spark.broadcast.Broadcast
+
+import scala.collection.mutable
+
+/**
+ * Process parameters trait, subclass must be independent of each other
+ */
+private[bigdl] trait ParameterProcessor
+  extends Serializable {
+  /**
+   * Collect global data according to operations list, usually executed in driver
+   *
+   * @param models cached models
+   * @param parameters [[AllReduceParameter]]
+   * @param metrics metrics
+   * @param state A table contained needed information
+   */
+  def collectGlobalData[T](models: RDD[Cache[T]],
+    parameters: AllReduceParameter[T],
+    metrics: Metrics,
+    state: Table)(implicit ev: TensorNumeric[T]) : Unit = {}
+
+  /**
+   * Advance operations to process parameters, usually executed in worker
+   *
+   * @param parameters [[AllReduceParameter]]
+   * @param state A table contained needed information
+   */
+  def processParameters[T](parameters: AllReduceParameter[T],
+    modelCache: Cache[T],
+    state: Table)(implicit ev: TensorNumeric[T]): Unit = {}
+
+  /**
+   * Advance operations to process parameters, usually executed in local optimer
+   *
+   * @param model the model to be trained
+   * @param state A table contained needed information
+   */
+  def processParameters[T](model: Module[T],
+    state: Table)(implicit ev: TensorNumeric[T]): Unit = {}
+}
+
+/**
+ * Process constant clipping
+ */
+private[bigdl] class ConstantClippingProcessor(min: Double, max: Double)
+  extends ParameterProcessor {
+  override def processParameters[T](parameters: AllReduceParameter[T],
+    modelCache: Cache[T],
+    state: Table)(implicit ev: TensorNumeric[T]): Unit = {
+    parameters.gradientPartition.clamp(min, max)
+  }
+
+  override def processParameters[T](model: Module[T],
+    state: Table)(implicit ev: TensorNumeric[T]): Unit = {
+    val gradients = model.getParameters()._2
+    gradients.clamp(min, max)
+  }
+}
+
+/**
+ * Process l2 norm clipping
+ */
+private[bigdl] class L2NormClippingProcessor(l2NormThreshold: Double)
+  extends ParameterProcessor {
+  override def collectGlobalData[T](models: RDD[Cache[T]],
+    parameters: AllReduceParameter[T],
+    metrics: Metrics,
+    state: Table)(implicit ev: TensorNumeric[T]) : Unit = {
+    val numFinishedModel = state.get[Int]("numFinishedModel").get
+    val parallelism = state.get[Int]("parallelism").get
+    val aggregatedG = state.get[Boolean]("aggregateG").get
+
+    val sumSquare = models.mapPartitions(modelIter => {
+      if (!aggregatedG) {
+        val getG = System.nanoTime()
+        parameters.aggregateGradientPartition(numFinishedModel)
+        metrics.add("aggregrateGradientParition average executor",
+          System.nanoTime() - getG)
+      }
+      val sum = Util.getSumsquareInParallel(parameters.gradientPartition, parallelism)
+      Iterator.single(sum)
+    }).reduce(_ + _)
+
+    state("aggregateG") = true
```

Comment:
should we mention the state aggregatedG is changed as a side affect in this method and the reason?